### DISCUSSION TITLE
What should be in the Vendure plugin template?

### DISCUSSION THREAD

**Discussion by martijnvdbrug** [vendurehq/vendure#2039] [Q&A]
Even though it was a great start, the original template is a bit outdated: https://github.com/vendure-ecommerce/plugin-template

I would like to update the template and make it more maintanable. To keep it maintainable, I propose to not include _every_ available feature in the template, because that means we need to keep all of those features up to date and in line with the docs.

So, what should at least be in the plugin? What I've got so far:

It should at least include:
* `example.plugin.ts`
* `example.service.ts`
* constants.ts with a loggerCtx and PLUGIN_OPTIONS symbol
* types.ts with an init options interface
* An options provider in the plugin metadata
* an init() static method on the plugin class and static options field.
* `test/e2e.spec`
* `test/dev-server.ts`

Up for discussion
:smile: graphql-codegen to generate Graphql types
:tada: Custom fields
:heart: Custom entities
:rocket: Admin UI + compilation
:eyes: shop/admin graphql extensions with a `resolver.ts`

Something else that's not on this list?

Let me know what you think.

## You can vote by commenting the corresponding emoticon to this message

**Comment by michaelbromley** [comment]
I'd like to also add that there are 2 use-cases for a plugin template:

1. Creating a new stand-alone plugin project with its own repo and tooling like codegen etc (this is how it is currently set up)
2. Adding a plugin to an existing Vendure app

I'd say that use-case 2 is much, much more common and we should optimize for this. Most Vendure plugins are internal to a single app. And even plugins created expressly for distribution via npm will likely be part of an existing monorepo.

So we should design any changes with this in mind. This makes the "codegen" option possibly moot, since we can usually assume the tooling for this already exists and is configured.

**Comment by martijnvdbrug** [comment]
Agreed, I also had option 2 in mind. 
In case of a monorepo, it's still useful to have codegen in place, so you can generate types per plugin. (At least, thats how I do it). 

In case of an existing project it doesn't make sense indeed.

I will leave it out, and add links in the Readme to stuff like codegen and Admin UI compilation

**Comment by tianyingchun** [comment]
1. it best to support typing for `mono` repo like customized field extension, like vscode `workspace` with yarn 3.0+

2. provides graphql `codegen`  recommend configurations (better seperate from other  plugin in `mono`)
